src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs

```diff
@@ -5025,12 +5025,62 @@ private static int WaitAnyCore(Task[] tasks, int millisecondsTimeout, Cancellati
 
         #region FromResult / FromException / FromCanceled
 
-        /// <summary>Creates a <see cref="Task{TResult}"/> that's completed successfully with the specified result.</summary>
+        /// <summary>Gets a <see cref="Task{TResult}"/> that's completed successfully with the specified result.</summary>
         /// <typeparam name="TResult">The type of the result returned by the task.</typeparam>
         /// <param name="result">The result to store into the completed task.</param>
         /// <returns>The successfully completed task.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)] // method looks long, but for a given TResult it results in a relatively small amount of asm
         public static Task<TResult> FromResult<TResult>(TResult result)
         {
+            // The goal of this function is to be give back a cached task if possible, or to otherwise give back a new task.
+            // To give back a cached task, we need to be able to evaluate the incoming result value, and we need to avoid as
+            // much overhead as possible when doing so, as this function is invoked as part of the return path from every async
+            // method and Task.FromResult. Most tasks won't be cached, and thus we need the checks for those that are to be as
+            // close to free as possible. This requires some trickiness given the lack of generic specialization in .NET.
+
+            if (null != (object?)default(TResult)) // help the JIT avoid the value type branches for ref types
```

Comment:
Can be replaced with `if (typeof(TResult).IsValueType)` 🙂  which is, I believe, looks better in terms of code readability

Reply:
> Can be replaced with if (typeof(TResult).IsValueType)

Are they both handled equally well/poorly in tier 0?

Reply:
@stephentoub just checked, pretty much the same, this condition won't be optimized to a const and branch won't be removed in tier0 for both cases. `IsValueType` adds some small overhead though (`call CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE`)

Reply:
Ok, thanks, seems reasonable to switch to it.  Another alternative is to remove the check entirely... I don't remember why it was added in the first place, and if the reason for that is still relevant a decade later....